CPLEX global scale

islerfab opened the pull request:
@beyeler 
This is a proposition for #16 . That way, all internal `.solve()` calls go through a singleton, which can take in any parameters that should be used as default beforehand. It's also possible to solve problems in parallel with non-default parameters by passing additional SolveParams. It may be required to propagate passing these additional parameters further up (e.g., to WinnerDetermination, ModelMIPs, etc.), but that way the basic logic is there (and setting the default parameters for all runs probably resolves #16 already).

beyeler commented on src/main/java/org/marketdesignresearch/mechlib/utils/CPLEXUtils.java:
```diff
@@ -20,10 +19,19 @@
+        solveParamMap.put(param, value);
+    }
+
     public void initializeSolveParams() {
         solveParamMap.put(SolveParam.CALC_DUALS, Boolean.TRUE);
         solveParamMap.put(SolveParam.LP_OPTIMIZATION_ALG, 2);
```
There are a lot of SolveParams set which are might too specific for a global setting. These were orginally only used to calculate the Norm MIPs and may should be moved to this location.

islerfab replied:
I agree!

islerfab commented:
@beyeler, Thanks for your input! I agree with what you're saying, I think I was not completely capturing the way you're planning to use it. I think I found a reasonable compromise, and refactored the CPLEXUtils accordingly. I documented all the methods, so I'm not writing the thoughts down here as well, please have a look inline. Thanks!

beyeler commented on src/main/java/org/marketdesignresearch/mechlib/utils/CPLEXUtils.java:
```diff
@@ -19,15 +18,55 @@
+     *  A helper function to initialize the solver with some reasonable default parameters.
+     *  Has to be actively called by the user.
+     */
     public void initializeSolveParams() {
+        solveParamMap.clear();
+        solveParamMap.put(SolveParam.DISPLAY_OUTPUT, log.isDebugEnabled());
+        solveParamMap.put(SolveParam.THREADS, 1);
+        solveParamMap.put(SolveParam.TIME_LIMIT, (double) TimeUnit.SECONDS.convert(1, TimeUnit.HOURS));
```
I agree that the DISPLAY_OUTPUT parameter would be a good choice for a "default" parameter. However the other two parameters I would not suggest to set as reasonalbe defaults. Setting threads to 1 is not a good choice if you have more cores and setting a time limit may prevent finding an optimal solution, even though you are really interested in an optimal solution.

beyeler commented on src/main/java/org/marketdesignresearch/mechlib/utils/CPLEXUtils.java:
```diff
@@ -19,15 +18,55 @@
+     * A helper function to initialize the solver with default parameters that were previously used in experiments
+     * of very specific CCG-price calculation MIPs
+     */
+    public void initializeNormSolveParams() {
+        initializeSolveParams();
```
Using the global parameters to set specific MIP parameters is not a good idea, because they stay active for the following solves which are might no norm problems. Therefore I suggest to either put these parameters to the location where the MIP is created or add the MIP as a parameter of this method. 

The latter has the advantage, that you could reuse these parameters for other problems and have a sort of reference what parameters should/can be set by a program.

beyeler replied:
see comment below.

beyeler commented:
Sorry. Just saw that this method initializeNormSolveParams is only used for the tests. So why not move them to the Test class? 
The other question would be if there should be some non default params set for the Payment and Norm Mips, or if we are fine as it is now, that users of the library just use default params.

islerfab commented:
I actually left the two methods for setting the parameters in there just as helpers, since they are used in some use cases. So I would use them in there. But I thought of the user setting the parameters by himself through the `setSolveParam` method, if he wants them to be applied for all future MIPs that are solved with the MechLib. The idea is that the user can still specify his MIP-specific parameters in each MIP, and the global parameters he has set are only applied on top of that (and if they were not overwritten on the MIP level).
I think it solves the issue you've created, but it may not be the most intuitive way. Any suggestions?

beyeler commented:
Yes now I understand your approach. I would suggest the following:

Because now JOpt has all reasonable defaults I would suggest to remove the two initializeSolveParams methods from the Utils class to the location where they are used.

The reason behind this is that setting parameters like the MARKOWITZ_TOLERANCE at a global scope is very bad practice in my opinion. There are some MIPs where I set this parameter on purpose and on the other MIPs I assume that this parameter is set to the default value, otherwise the code might break (e.g. for the Approximate linear price calculation in MLCA with bounds).

The only parameter I can think of at the moment that should be allowed to set on a global scope is really the Thread count, because you need to tune this parameter to allow to execute multiple instances in paralell on the same cluster node. I know that as soon as we start using the deterministic timelimit things get more complicated because the time limit also depends on the number of threads used. I do not have a solution to this yet.

islerfab commented:
We can talk about this in our call, but I think having these methods in the wrapper is the most appropriate place. A user is not expected to call them, the primary use case of the wrapper is to set the parameters via `setSolveParam(...)`. The methods only represent a previously used setting, and the user is free to set this as a starting point. But in your case, you wouldn't do this, you would probably only set the thread count globally, and then set the markowitz tolerance on a per-mip basis.

I updated the documentation and changed the implementation of the method to literally only use wrapper-internal method, to make clear that this is only a shortcut, and that the point of the wrapper is to set global parameters via the `setSolveParam(...)` method. Do you agree with this, or would you still do it differently?
